modules/dnn/src/cuda/permute.cu

```diff
@@ -0,0 +1,152 @@
+    }
+
+    template <class T>
+    void permute(
+        const Stream& stream,
+        TensorSpan<T> output, TensorView<T> input,
+        std::vector<std::size_t> order)
+    {
+        CV_Assert(output.rank() == input.rank());
+        CV_Assert(input.rank() == order.size());
+        CV_Assert(input.size() == output.size());
+
+        /* if there are axes in the begining that can be squeezed out and are not permuted,
+         * we can eliminate those axes
+         *
+         * Reasoning:
+         * ----------
+         * Suppose an item's indices in the input tensor be [i1, i2, ...]. The indices in the
+         * output tensor will be some permutation of the input tensor indices. Let the output
+         * tensor indices be [o1, o2, ...]. The permutation operation essentially copies items
+         * from the input tensor to new locations in the output tensor as dictated by the indices.
+         *
+         * If the size of the first axis of the input and output tensor is one and these axes are
+         * not involved in any permutation, i.e. order[0] = 0, the input and output indicies for
+         * all the elements will be of the form be [0, i2, ...] and [0, o2, ...] respectively.
+         * The first index does not contribute to the element's address calculation and hence does
+         * nothing other than consuming few cycles.
+         */
+        while (order[0] == 0 && input.get_axis_size(0) == 1 && output.get_axis_size(0) == 1) {
+            /* remove the axes */
+            input.squeeze(0);
+            output.squeeze(0);
+
+            /* when we remove axis zero, the axis index will be one less than the previous index
+             * for all the remaining axes
+             */
+            order.erase(order.begin());
+            for (auto& axis : order)
+                axis--;
+
+            /* optimizations should not break the invariants for the permutation operation */
+            CV_Assert(output.rank() == input.rank());
+            CV_Assert(input.rank() == order.size());
+            CV_Assert(input.size() == output.size());
+        }
+
+        auto rank = output.rank();
+        auto inShape = input.shape_as_vector();
+        auto outShape = output.shape_as_vector();
+
+        std::vector<std::size_t> inStride(rank), outStride(rank);
+        inStride.back() = 1;
+        outStride.back() = 1;
+        /* garbage, ..., garbage, 1 */
+
+        std::copy(std::begin(inShape) + 1, std::end(inShape), std::begin(inStride));
+        std::copy(std::begin(outShape) + 1, std::end(outShape), std::begin(outStride));
+        /* dim[0], dim[1], ..., dim[-1], 1 */
+
+        std::partial_sum(inStride.rbegin(), inStride.rend(), inStride.rbegin(), std::multiplies<std::size_t>());
+        std::partial_sum(outStride.rbegin(), outStride.rend(), outStride.rbegin(), std::multiplies<std::size_t>());
+        /* stride[0], stride[1], ..., stride[-2], 1 */
+
+        CV_Assert(rank >= 2 && rank <= 5);
```

Comment:
The maximum tensor rank is defined by `CSL_MAX_TENSOR_RANK` (which is five currently). The maximum tensor rank has been hardcoded here (and in many other kernel launch code).

The following structure frequently arises in many places (permute, slice, concat, etc):
```
if (rank == 5) {
    launch_some_kernel<T, 5>(args...);
} else if (rank == 4) {
     launch_some_kernel<T, 4>(args...);
} else if (rank == 3) {
     launch_some_kernel<T, 3>(args...);
} else if (rank == 2) {
     launch_some_kernel<T, 2>(args...);
}
```

It's not very straightforward how the maximum tensor rank limit could be incorporated here. I can think of recursive-template-based solution specific to each kernel launch code. I am not quite sure how to write a neat generic version which can be used at multiple places and be placed in the `cuda4dnn/cxx_utils` directory.

Reply:
Do you have an example of this?  Why does the tensor's rank need to be a compile time constant like this?

You can make this kind of thing work, that is, mapping a small runtime integer value to a compile time integer value using template recursion with `if (compile_time_idx == runtime_idx) do_thing<compile_time_idx>()` inside the recursion.  You would have to limit the depth to a known max, but that's not a problem in this case.  I've done this kind of thing in many instances and it works just fine.  But I still wonder if this whole thing couldn't be avoided?

Reply:
permute kernel: https://github.com/opencv/opencv/blob/d4a04f8486b98ec18b5eaf40a1ac17bbaacd51a1/modules/dnn/src/cuda/permute.cu#L30-L49

`launch_permute_kernel`: https://github.com/opencv/opencv/blob/d4a04f8486b98ec18b5eaf40a1ac17bbaacd51a1/modules/dnn/src/cuda/permute.cu#L52-L73

The `launch_permute_kernel` is only there to prevent code duplication.

The permute kernel takes two arrays of tensor strides and an array containing the permute order. The sizes of all of the arrays are equal to the rank. In total there are 15 numbers in the worst case (rank 5 tensors). There are two pointers, stride loop counter, index loop counter and two indices in addition to the intermediate registers required for computation.

**Profiling results:**
- as a template parameter (rank 4):
  - Register count: 29
  - Time: 332.22us

- as an argument (rank 4):
  - Register count: 31
  - Time: 758.40us

- as a template parameter (rank 3):
  - Register count: 26
  - Time: 33.54us

- as an argument (rank 3):
  - Register count: 31
  - Time: 74.37us

Note: The kernel launched by the same layer was used for all four measurements. A batch size of 10 was used to execute rank 4 permute and a single image was used for rank 3 permute. Hence, there is a factor of 10 in the timings.

The PTX shows that the argument version makes heavy use of local memory whereas the template does not which is presumably the cause for the difference.

I have made the rank as a template parameter so that the for small rank tensors, the strides and the order can be cached in the registers. This was even more important before the int64 to int32 index type optimization was made.

---

Yes, `if (compile_time_idx == runtime_idx) do_thing<compile_time_idx>()` will work but I will have to write a recursive template for every kernel launch code. I am not sure how `do_thing` could be passed to a generic version (same TMP code for all `do_thing` like functions).

I have a preprocessor based solution in mind but I think preprocessor should be the last resort. 

I have the following in mind:
1. write a separate recursive template for each such function
  - use the preprocessor to generate the recursive template for each such function
2. wrap the function inside a class and write a general TMP code which works for all such functions

Reply:
Nice, those are great speed improvements :)

Yeah, I'm not sure how to do this without doing one of the things you suggest.  It's too bad you can't pass function templates as template template arguments in C++, as that would do the trick. I would go with the macro option.